### PR TITLE
Osemgrep fingerprint

### PR DESCRIPTION
**Pull request by reynir**
This aligns how fingerprints are computed with pysemgrep.

First, `pattern-sanitizers` is omitted. This is done in pysemgrep due to `"pattern-sanitizers"` is not in `RuleValidation.PATTERN_KEYS`.
https://github.com/semgrep/semgrep/blob/53caa62f16ed9e585f74d583dcce9a438dbe14b6/cli/src/semgrep/rule.py#L307
https://github.com/semgrep/semgrep/blob/53caa62f16ed9e585f74d583dcce9a438dbe14b6/cli/src/semgrep/rule_lang.py#L459-L472

Second, the sort order of the patterns was incorrect. It is not enough to flatten the patterns and then sort them.
```yaml
- patterns:
  - pattern-either:
    - pattern: a
    - pattern: c
  - pattern: b
```
The above patterns should result in `a c b` as `a c` comes before `b`. However, flattening the patterns we get `a b c`.

The first item seems to me like a bug in pysemgrep(?), but fixing it makes a new test pass. The second item fixes the output but does not make any new tests pass as the sort order is still different.

### CODE REVIEW

**Comment by reynir** [comment]
Sorting the results in pysemgrep --sarif, similar to #9636, makes two new tests pass with osemgrep. Alternatively, we can rewrite Sarif_output.ml and caller so we don't sort the results in osemgrep.

**Review by aryx — COMMENTED** [review]
lgtm.
I'm a bit worried about the change of fingerprints, and whether this can have an impact on the semgrep WebApp, but maybe @ajbt200128 know.

File: cli/src/semgrep/formatter/sarif.py [hunk 1]
```diff
@@ -414,6 +414,7 @@ def format(
         - Full specification is at:
             https://docs.oasis-open.org/sarif/sarif/v2.1.0/cs01/sarif-v2.1.0-cs01.html
         """
+        sorted_findings = sorted(rule_matches)
```
Comment (ajbt200128) [inline cli/src/semgrep/formatter/sarif.py:417]: this seems fine, but curious why we added it (and maybe leave a comment explaining why)

File: cli/tests/e2e/snapshots/test_output_sarif/test_sarif_output_with_source/results.sarif [hunk 2]
```diff
@@ -11,61 +11,61 @@
       "results": [
         {
           "fingerprints": {
-            "matchBasedId/v1": "62b4a09c4569768898c43c09fa0a5b95b7e93257ef3a0911a5c379b6265b4d49fa4aecd5782461632e9aef4779af02d7cad4405b9a5318a0e5ffe9a5bd8daeae_0"
+            "matchBasedId/v1": "33c7ad418bcb7f83d9dcec68b2a8aa78ace93efbc20a12297ea7e15594ce23f5bca80b0958952b14dad3e874370c9ca7f991d2e1414adc33d243f133b1ff2811_0"
```
Comment (ajbt200128) [inline cli/tests/e2e/snapshots/test_output_sarif/test_sarif_output_with_source/results.sarif:14]: @aryx it looks like the fingerprints just moved order in the json, but didn't change, if this was what you're talking about

**Review by ajbt200128 — APPROVED** [review]
This looks good. I'm ok changing osemgrep fingerprints to match pysemgrep. It shouldn't be noisy, as I don't think anyone is using osemgrep in CI

File: cli/src/semgrep/formatter/sarif.py (hunk 1 above)
Comment (reynir) [reply cli/src/semgrep/formatter/sarif.py:417]: In osemgrep we use the sorted json output to build the sarif output. IIRC there is a comment somewhere that this is done so the output is deterministic in the face of parallelism. I can add a comment in the code on Monday (anyone feel free to do so before then). It should be possible to rewrite it to not use the json output in osemgrep and thus not sort.

**Pull request merged**